### PR TITLE
feat(cf): Implement on-demand caching

### PR DESCRIPTION
**Pull request by stuart-pollock**
- for performance reasons, we introduced the caching of individual
CloudFoundryServerGroup objects

spinnaker/spinnaker#4083

### CODE REVIEW

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Applications.java [hunk 1]
```diff
@@ -88,6 +117,14 @@ public CloudFoundryServerGroup findById(String guid) {
       .collect(toList());
   }
 
+  @Nullable
+  public CloudFoundryServerGroup findServerGroupByNameAndSpaceId(String name, String spaceId) {
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Applications.java]: `findServerGroupId` hits API, but `findById` uses the cache. Is there potential drift between them? If we are going to limit what `findServerGroupByNameAndSpaceId` can find to those things in the cache, can we instead just iterate over the cache looking for this name/space combination? And a view on this cache if you think this is called frequently enough to optimize?

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Organizations.java [hunk 2]
```diff
@@ -70,4 +71,35 @@ public CloudFoundryOrganization findById(String orgId) throws CloudFoundryApiExc
...
+    // fully populates the space guid which is what Cloud Foundry's API expects as an input, not the name.
+    Optional<CloudFoundryOrganization> organization = findByName(space.getOrganization().getName());
+
+    Optional<CloudFoundrySpace> spaceOptional = organization
+      .map(org -> findSpaceByName(org, space.getName()));
+
+    spaceOptional.ifPresent(spaceCase -> {
+      if (!(space.getName().equals(spaceCase.getName()) &&
+        space.getOrganization().getName().equals(spaceCase.getOrganization().getName()))) {
+        throw new CloudFoundryApiException("Org or Space name not in correct case");
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Organizations.java]: Is this possible? Is CAPI's name filter case insensitive? (https://apidocs.cloudfoundry.org/7.9.0/organizations/list_all_organizations.html)

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/provider/CloudFoundryProvider.java [hunk 3]
```diff
@@ -50,7 +50,7 @@
     singletonMap(new SearchableResource(APPLICATIONS.getNs(), "cloudfoundry"), new ApplicationSearchResultHydrator());
 
   private final Map<String, String> urlMappingTemplates = emptyMap();
-  private final String id = "cloudfoundry";
+  public static final String PROVIDER_ID = "cloudfoundry";
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/provider/CloudFoundryProvider.java]: I believe this is a breaking change somehow, that there are places that rely on `getId()` on the provider. Notice the `@Getter` on the type.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Organizations.java (hunk 2 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/client/Organizations.java]: Yes, the CAPI name filter is case-insensitive.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/provider/CloudFoundryProvider.java (hunk 3 above)
Comment (stuart-pollock) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/provider/CloudFoundryProvider.java]: If `CloudFoundryProvider` implemented `CloudProvider`, then there might be a problem.  It implements `SearchableProvider`, which does not have that method.

**Pull request merged**